### PR TITLE
WIP: runtime changes for static interface method support

### PR DESCRIPTION
**Pull request by trylek**
Apparently I still don't have push rights to your fork so for now I'm pushing the commits to my fork. I originally tried to create this PR against your version of the branch but GitHub doesn't let me select it as the base repo, probably due to the fact that I'm not authorized to push. I assume this is a technical detail we'll easily resolve once you're back from vacation; for now I believe this WIP PR to be a useful vehicle for me to publish and comment on my progress and for you to review it when you're back in office.

Thanks

Tomas

### CODE REVIEW

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 1]
```diff
@@ -2943,10 +2946,6 @@ MethodTableBuilder::EnumerateClassMethods()
                 {
                     BuildMethodTableThrowException(BFA_VIRTUAL_PINVOKE_METHOD);
                 }
-                if(IsMdStatic(dwMemberAttrs))
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp:2946]: We should not permit static virtual methods on classes/structs. By our ecma design, those should continue to throw.

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 2]
```diff
@@ -4770,17 +4769,15 @@ VOID MethodTableBuilder::TestMethodImpl(
...
         BuildMethodTableThrowException(IDS_CLASSLOAD_MI_NONVIRTUAL_DECL);
     }
-    if (!IsMdVirtual(dwImplAttrs))
+    if (!IsMdVirtual(dwImplAttrs) && !IsMdStatic(dwImplAttrs))
     {
+        // TODO: better error message? When non-virtual, it must be static.
         BuildMethodTableThrowException(IDS_CLASSLOAD_MI_MUSTBEVIRTUAL);
     }
     // Virtual methods cannot be static
-    if (IsMdStatic(dwDeclAttrs))
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp:4778]: We should not permit static virtual methods on classes/structs. By our ecma design, those should continue to throw.

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 3]
```diff
@@ -5023,15 +5020,11 @@ MethodTableBuilder::ValidateMethods()
         {
             if (!IsMdVirtual(it.Attrs()))
             {   // Non-virtual methods cannot participate in a methodImpl pair.
+                /* TODO: we should probably check that we're overriding a virtual static method here
                 BuildMethodTableThrowException(IDS_CLASSLOAD_MI_MUSTBEVIRTUAL, it.Token());
+                */
             }
         }
-
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp:5029]: We should not permit static virtual methods on classes/structs. By our ecma design, those should continue to throw.

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 4]
```diff
@@ -5613,15 +5606,6 @@ MethodTableBuilder::ProcessMethodImpls()
     DeclaredMethodIterator it(*this);
     while (it.Next())
     {
-        // Non-virtual methods cannot be classified as methodImpl - we should have thrown an
-        // error before reaching this point.
-        CONSISTENCY_CHECK(!(!IsMdVirtual(it.Attrs()) && it.IsMethodImpl()));
-
-        if (!IsMdVirtual(it.Attrs()))
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp:5620]: This should only be allowed for static methods

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 5]
```diff
@@ -2376,6 +2377,7 @@ MethodTableBuilder::EnumerateMethodImpls()
                 {
                     BuildMethodTableThrowException(IDS_CLASSLOAD_MI_ILLEGAL_TOKEN_BODY);
                 }
+                /* TODO: what to do about this check
```
Comment (davidwrighton) [inline src/coreclr/vm/methodtablebuilder.cpp:2380]: The spec update describes the changes here. This check should be preserved for instance virtuals, but there should be a base check for static virtual overrides.

File: src/coreclr/vm/clsload.cpp [hunk 6]
```diff
@@ -2958,6 +2958,25 @@ ClassLoader::LoadApproxTypeThrowing(
         CorElementType type = ELEMENT_TYPE_END;
         IfFailThrowBF(sigptr.GetElemType(&type), BFA_BAD_SIGNATURE, pModule);
 
+        // TODO: For some reason, the base class BaseScenario1 - DerivedScenario1 inheritance
```
Comment (davidwrighton) [inline src/coreclr/vm/clsload.cpp:2961]: This is odd. Possibly we just have a ilasm bug.

File: src/coreclr/vm/contractimpl.cpp [hunk 7]
```diff
@@ -434,9 +434,11 @@ DispatchMap::CreateEncodedMapping(
             curTargetSlot = -1;
             do
             {
+                /* TODO: right now it's unclear to me whether the SVMs should go in the method table or not, for now I assume they should.
```
Comment (davidwrighton) [inline src/coreclr/vm/contractimpl.cpp:437]: My initial thought was to store this sort of detail in a parallel table to the existing scheme, but we'll want to take a close look here to see if this is the right approach.

File: src/coreclr/vm/prestub.cpp [hunk 8]
```diff
@@ -1950,7 +1950,7 @@ extern "C" PCODE STDCALL PreStubWorker(TransitionBlock* pTransitionBlock, Method
         CONSISTENCY_CHECK(GetAppDomain()->CheckCanExecuteManagedCode(pMD));
 
         MethodTable* pDispatchingMT = NULL;
-        if (pMD->IsVtableMethod())
+        if (!pMD->IsStatic() && pMD->IsVtableMethod())
```
Comment (davidwrighton) [inline src/coreclr/vm/prestub.cpp:1953]: I don't like the idea that static methods would go in the vtable.

**Review by davidwrighton — CHANGES_REQUESTED** [review]
I have a set of comments on the change in progress, although I'm really happy to see at least 1 test pass. Very exciting.

**Comment by trylek** [comment]
Closing in favor of a new PR against David's private branch.

**Pull request closed without merging**